enhancements/machine-config/os-images-streams.md

```diff
@@ -0,0 +1,1075 @@
+- **Backward compatibility**: Existing clusters continue working; streams are opt-in via feature gate
+- **Multi-source stream configuration**: Support CLI arguments, release ImageStream, and ConfigMap sources with defined precedence
+
+### Non-Goals
+
+- **Supporting unlimited concurrent OS streams**: While the architecture supports
+multiple OS streams, this initial implementation focuses on enabling 2-3
+concurrent streams (e.g., RHCOS 9 + RHCOS 10, or RHCOS 10 standard + minimal).
+Supporting large numbers of concurrent streams with complex version matrices is
+not a goal for the initial releases.
+
+- **HCP/Hypershift architecture support**: Supporting OS streams in
+HCP/Hypershift environments (which use NodePools instead of MachinConfigPools)
+is deferred to future work. See [Topology Considerations](#topology-considerations)
+for details.
+
+- **Automatic migration or upgrade paths**: The enhancement does not include
+automated migration logic or upgrade orchestration. Administrators must manually
+select streams for their pools.
+
+- **Bidirectional stream switching**: Only RHEL 9 → RHEL 10 migration is supported
+in the initial Tech Preview implementation. RHEL 10 → RHEL 9 downgrade is not
+supported and may be considered for future releases. The primary use case is
+forward migration to newer OS versions.
+
+- **Rollback from failed stream changes**: If a stream migration fails mid-update
+(e.g., node fails to boot with new OS), automated rollback to the previous stream
+is not guaranteed in Tech Preview. Manual recovery procedures may be required.
+Standard MachineConfigPool rollback mechanisms apply, but stream-specific rollback
+is out of scope for the initial release.
+
+- **Version skew enforcement**: Enforcing compatibility rules between different
+OS versions and OpenShift platform versions is not included in this enhancement.
+This functionality is planned for OpenShift 4.22.
+
+- **Boot image management**: Updating boot images or installation media to match
+the selected stream is not covered by this enhancement.
+
+- **bootc integration**: This enhancement continues to use rpm-ostree for image
+updates in OpenShift 4.x. Migration to bootc-based updates is planned for
+OpenShift 5.0+.
+
+- **Changing the update mechanism**: The underlying OS update mechanism
+(rpm-ostree) remains unchanged. This enhancement only provides stream selection
+capability.
+
+## Proposal
+
+To implement the functionality this enhancement provides, some changes are
+required in the MCO, the released images payload, and the CoreOS images.
+The following sections describe all the required changes.
+
+### Implementation Status
+
+This enhancement is being delivered incrementally across multiple releases. Features marked as **"(Planned)"** or **"TBD"** indicate functionality that is designed but not yet fully implemented.
+
+#### Completed (4.21 Dev Preview)
+- API definitions: `osImageStream` field in MachineConfigPool, OSImageStream v1alpha1 resource
```

Comment:
This has not yet merged. Perhaps we don't need to be this granular since we don't change the enhancement that often, so status will easily become stale. Maybe we can just describe the implementation and rough timeline, and not the details in this manner.